File: scripts/item-generator/catalog/collections.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Ian Lucas. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { CS2ContainerType, CS2ItemType } from "../../../src/economy-types.ts";
import { assert } from "../../../src/utils.ts";
import { type ItemGeneratorContext } from "../types.ts";
import { tryAddTranslation } from "./translations.ts";

export function getCollection(
    ctx: ItemGeneratorContext,
    itemId: number,
    collection?: string
): { collection: string | undefined; collectionImage: string | undefined } {
    let collectionImage: string | undefined;
    if (collection !== undefined) {
        const itemSet = ctx.gameItems.item_sets[collection];
        assert(itemSet);
        assert(itemSet.name);
        tryAddTranslation(ctx, itemId, "collectionName", itemSet.name);
        tryAddTranslation(ctx, itemId, "collectionDesc", itemSet.set_description);
        collectionImage = ctx.itemSetImage[collection];
    }
    return { collection, collectionImage };
}

export function getItemCollection(
    ctx: ItemGeneratorContext,
    itemId: number,
    itemKey: string
): { collection: string | undefined; collectionImage: string | undefined } {
    return getCollection(ctx, itemId, ctx.itemSetItemKey[itemKey]);
}

export function addContainerItem(ctx: ItemGeneratorContext, itemKey: string, id: number): void {
    if (!ctx.containerItems.has(itemKey)) {
        ctx.containerItems.set(itemKey, id);
    }
}

export function getClientLootListItems(
    ctx: ItemGeneratorContext,
    clientLootListKey: string,
    items: string[] = []
): string[] {
    if (!ctx.gameItems.client_loot_lists[clientLootListKey]) {
        return [];
    }
    for (const itemOrClientLootListKey of Object.keys(ctx.gameItems.client_loot_lists[clientLootListKey])) {
        if (ctx.containerItems.has(itemOrClientLootListKey)) {
            items.push(itemOrClientLootListKey);
        } else {
            getClientLootListItems(ctx, itemOrClientLootListKey, items);
        }
    }
    return items;
}

export function getContainerType(name?: string, type?: CS2ItemType): CS2ContainerType | undefined {
    switch (true) {
        case name?.includes("Souvenir"):
            return CS2ContainerType.SouvenirCase;
        case type === CS2ItemType.Weapon:
            return CS2ContainerType.WeaponCase;
        case type === CS2ItemType.Sticker:
            return CS2ContainerType.StickerCapsule;
        case type === CS2ItemType.Graffiti:
            return CS2ContainerType.GraffitiBox;
        default:
            return undefined;
    }
}

File: scripts/item-generator-cs/Catalog/Collections.cs
namespace ItemGenerator.Catalog;

public static class Collections
{
    public static (string? Collection, string? CollectionImage) GetCollection(
        ItemGeneratorContext ctx, int itemId, string? collection)
    {
        string? collectionImage = null;
        if (collection == null) return (null, null);

        var itemSets = KvHelper.GetChild(ctx.GameItems!, "item_sets");
        var itemSet = KvHelper.GetChild(itemSets, collection);
        if (itemSet == null) return (collection, null);

        var name = KvHelper.GetString(itemSet, "name");
        var setDescription = KvHelper.GetString(itemSet, "set_description");
        Translations.TryAddTranslation(ctx, itemId, "collectionName", name);
        Translations.TryAddTranslation(ctx, itemId, "collectionDesc", setDescription);

        if (ctx.ItemSetImage.TryGetValue(collection, out collectionImage))
            return (collection, collectionImage);

        return (collection, null);
    }

    public static (string? Collection, string? CollectionImage) GetItemCollection(
        ItemGeneratorContext ctx, int itemId, string itemKey)
    {
        ctx.ItemSetItemKey.TryGetValue(itemKey, out var collection);
        return GetCollection(ctx, itemId, collection);
    }

    public static void AddContainerItem(ItemGeneratorContext ctx, string itemKey, int id)
    {
        ctx.ContainerItems.TryAdd(itemKey, id);
    }

    public static List<string> GetClientLootListItems(ItemGeneratorContext ctx, string clientLootListKey, List<string>? items = null)
    {
        items ??= [];
        var clientLootLists = KvHelper.GetChild(ctx.GameItems!, "client_loot_lists");
        var lootList = KvHelper.GetChild(clientLootLists, clientLootListKey);
        if (lootList == null) return items;

        foreach (var child in lootList)
        {
            var key = child.Key;
            if (ctx.ContainerItems.ContainsKey(key))
                items.Add(key);
            else
                GetClientLootListItems(ctx, key, items);
        }

        return items;
    }

    public static int? GetContainerType(string? name, string? type)
    {
        if (name?.Contains("Souvenir") == true) return CS2ContainerType.SouvenirCase;
        if (type == CS2ItemType.Weapon) return CS2ContainerType.WeaponCase;
        if (type == CS2ItemType.Sticker) return CS2ContainerType.StickerCapsule;
        if (type == CS2ItemType.Graffiti) return CS2ContainerType.GraffitiBox;
        return null;
    }
}
